src/runtime_src/core/edge/drm/zocl/sched_exec.c

```diff
@@ -3288,6 +3292,68 @@ static irqreturn_t sched_cq_isr(int irq, void *arg)
 	return IRQ_HANDLED;
 }
 
+static int watchdog_thread(void *data)
+{
+	struct drm_zocl_dev *zdev = data;
+	struct zocl_watchdog_dev *watchdog = zdev->watchdog;
+
+	if (!watchdog)
+		goto exit;
+
+	watchdog->ops->init(watchdog);
+	while (!kthread_should_stop()) {
+		struct watchdog_cfg cfg = {
+			.skd_run = false,
+			.cmc_run = false,
+			.cq_thread_run = false,
+			.sched_thread_run = false,
+		} ;
+		struct task_struct *tsk;
+
+		rcu_read_lock();
+		for_each_process(tsk) {
+			if (!strcmp(tsk->comm, "skd")) {
+				/*DRM_INFO("skd(pid: %d) is running\n",
+					task_pid_nr(tsk));*/
+				cfg.skd_run = true;
+			}
+			if (!strncmp(tsk->comm, CMC, strlen(CMC))) {
+				/*DRM_INFO("cmc(pid: %d) is running\n",
+					task_pid_nr(tsk));*/
+				cfg.cmc_run = true;
+			}
+		}
+		rcu_read_unlock();
+
+		/*
+		 * Notes:
+		 * 1. We don't expect the sched thread and cq thread exit, so
+		 *    we don't need a lock when checking the thread state.
+		 * 2. Other than TASK_DEAD, what else state should be monitor?
+		 */
+		if (zdev->exec && zdev->exec->cq_thread &&
+			zdev->exec->cq_thread->state != TASK_DEAD)
+			cfg.cq_thread_run = true;
+		/* DRM_WARN("ert kthread state: %lx\n",
+			zdev->exec->cq_thread->state);*/
+
+		if (g_sched0.sched_thread &&
+			g_sched0.sched_thread->state != TASK_DEAD)
+			cfg.sched_thread_run = true;
+		/*DRM_WARN("sched kthread state: %lx\n",
+			g_sched0.sched_thread->state);*/
+
+		watchdog->ops->config(watchdog, cfg);
+		msleep(ZOCL_WATCHDOG_FREQ);
+		schedule();
+	}
+	watchdog->ops->fini(watchdog);
+exit:
+	zdev->exec->watchdog_thread = NULL;
```

Comment:
I am not sure if this will really take effect? given you are calling
exec_core->watchdog_thread = kthread_run(watchdog_thread...)
On VCK5000, where zdev->watchdog = NULL, so which line will be executing first?

Reply:
as discussed offline, make it better to check whether watchdog is available before launching the thread